pkg/alignchecker/alignchecker.go

```diff
@@ -14,152 +14,239 @@
-		goStruct: reflect.TypeOf(proxymap.Proxy4Key{}),
-	},
-	reflect.TypeOf(C.struct_proxy4_tbl_value{}): {
-		sizeOfC:  C.sizeof_struct_proxy4_tbl_value,
-		goStruct: reflect.TypeOf(proxymap.Proxy4Value{}),
-	},
-	reflect.TypeOf(C.struct_proxy6_tbl_key{}): {
-		sizeOfC:  C.sizeof_struct_proxy6_tbl_key,
-		goStruct: reflect.TypeOf(proxymap.Proxy6Key{}),
-	},
-	reflect.TypeOf(C.struct_proxy6_tbl_value{}): {
-		sizeOfC:  C.sizeof_struct_proxy6_tbl_value,
-		goStruct: reflect.TypeOf(proxymap.Proxy6Value{}),
-	},
-	reflect.TypeOf(C.struct_sock_key{}): {
-		sizeOfC:  C.sizeof_struct_sock_key,
-		goStruct: reflect.TypeOf(sockmap.SockmapKey{}),
-	},
-	reflect.TypeOf(C.struct_ep_config{}): {
-		sizeOfC:  C.sizeof_struct_ep_config,
-		goStruct: reflect.TypeOf(configmap.EndpointConfig{}),
-	},
+func getStructInfosFromDWARF(d *dwarf.Data, toCheck map[string]reflect.Type) (map[string]structInfo, error) {
+	structs := make(map[string]structInfo)
+
+	r := d.Reader()
+
+	for entry, err := r.Next(); entry != nil && err == nil; entry, err = r.Next() {
+		// Read only DWARF struct entries
+		if entry.Tag != dwarf.TagStructType {
+			continue
+		}
+
+		t, err := d.Type(entry.Offset)
+		if err != nil {
+			return nil, fmt.Errorf("cannot read DWARF info section at offset %d: %s",
+				entry.Offset, err)
+		}
+
+		st := t.(*dwarf.StructType)
+
+		if _, found := toCheck[st.StructName]; found {
+			unionCount := 0
+			offsets := make(map[string]int64)
+			for _, field := range st.Field {
+				n := field.Name
+				// Create surrogate names ($union0, $union1, etc) for unnamed
+				// union members
+				if n == "" {
+					if t, ok := field.Type.(*dwarf.StructType); ok {
+						if t.Kind == "union" {
+							n = fmt.Sprintf("$union%d", unionCount)
+							unionCount++
+						}
+					}
+				}
+				offsets[n] = field.ByteOffset
+			}
+			structs[st.StructName] = structInfo{
+				size:         st.ByteSize,
+				fieldOffsets: offsets,
+			}
+		}
+	}
+
+	return structs, nil
+}
+
+func check(toCheck map[string]reflect.Type, structs map[string]structInfo) error {
+	for name, g := range toCheck {
+		c, found := structs[name]
+		if !found {
+			return fmt.Errorf("C struct %s not found", name)
+		}
+
+		if c.size != int64(g.Size()) {
+			return fmt.Errorf("struct sizes do not match: %s (%d) vs %s (%d)",
+				g, g.Size(), name, c.size)
+		}
+
+		for i := 0; i < g.NumField(); i++ {
+			fieldName := g.Field(i).Tag.Get("align")
+			// Ignore fields without `align` struct tag
+			if fieldName == "" {
+				continue
+			}
+			goOffset := int64(g.Field(i).Offset)
+			cOffset := structs[name].fieldOffsets[fieldName]
+			if goOffset != cOffset {
+				return fmt.Errorf("%s.%s offset (%d) does not match with %s.%s (%d)",
+					g, g.Field(i).Name, goOffset, name, fieldName, cOffset)
+			}
+		}
+	}
+
+	return nil
 }
 
-func init() {
-	for k, v := range cToGO {
-		err := compareStructs(k, v)
+// Adopted from elf.File.DWARF
+// https://github.com/golang/go/blob/master/src/debug/elf/file.go (go1.11.2).
```

Comment:
Per the [LICENSE](https://github.com/golang/go/blob/master/LICENSE), I think we need to retain the license/copyright notice for the below code, although we have changed it. You might consider keeping the below in a separate file in this package to make it explicit and clear.

Reply:
> Per the [LICENSE](https://github.com/golang/go/blob/master/LICENSE), I think we need to retain the license/copyright notice for the below code

Oops, you are right: `<..> **with or without
modification**, are permitted provided that the following conditions are
met: <..>`. I will move to a separate file and add the license.